conda-store-server/conda_store_server/action/download_packages.py

```diff
@@ -28,14 +43,149 @@ def action_fetch_and_extract_conda_packages(
             file_path = pkgs_dir / filename
             count_message = f"{packages_searched} of {total_packages}"
             with filelock.FileLock(str(lock_filename)):
```

Comment:
I can't comment on the lines above, but I've seen instances where `package['url']` is None. You also have `fn` (filename) as a metadata field I think? In case you want to provide a fallback.

Reply:
I'll keep this in mind, but I suggest we keep this as is for now. I was trying to copy the upstream code as close as possible and in `package_cache_data.py` it's always just the url:

```py
            url = self._urls_data.get_url(package_filename)
            package_cache_record = PackageCacheRecord.from_objects(
                raw_json_record,
                url=url,
                fn=basename(package_tarball_full_path),
                md5=md5,
                size=getsize(package_tarball_full_path),
                package_tarball_full_path=package_tarball_full_path,
                extracted_package_dir=extracted_package_dir,
            )

```